crates/astria-sequencer/src/accounts/action.rs

```diff
@@ -25,56 +26,74 @@ use crate::{
 /// Fee charged for a `Transfer` action.
 pub(crate) const TRANSFER_FEE: u128 = 12;
 
+pub(crate) async fn transfer_check_stateful<S: StateReadExt + 'static>(
+    action: &TransferAction,
+    state: &S,
+    from: Address,
+) -> Result<()> {
+    ensure!(
+        state.is_allowed_fee_asset(action.fee_asset_id).await?,
+        "invalid fee asset",
+    );
+
+    let transfer_asset_id = action.asset_id;
+
+    let from_fee_balance = state
+        .get_account_balance(from, action.fee_asset_id)
+        .await
+        .context("failed getting `from` account balance for fee payment")?;
+
+    // if fee asset is same as transfer asset, ensure accounts has enough funds
+    // to cover both the fee and the amount transferred
+    if action.fee_asset_id == transfer_asset_id {
+        let payment_amount = action
+            .amount
+            .checked_add(TRANSFER_FEE)
+            .ok_or(anyhow!("transfer amount plus fee overflowed"))?;
```

Comment:
no need to `ok_or(anyhow!`:

```suggestion
            .context("transfer amount plus fee overflowed")?;
```